cmd/verify/verify.go

```diff
@@ -40,40 +55,110 @@ func main() {
+				// Send the errorus message to an error queue so it can be analyzed.
+				if err := broker.SendMessage(mq, d.CorrelationId, config.Broker.Exchange, config.Broker.RoutingError, d.Body); err != nil {
+					log.Error("faild to publish message, reason: ", err)
+				}
+				// Restart on new message
+				continue
+			} else {
+				e := reflect.ValueOf(&m).Elem()
+				for i := 0; i < e.NumField(); i++ {
+					if e.Field(i).Interface() == "" || e.Field(i).Interface() == 0 {
+						log.Errorf("%s is missing", e.Type().Field(i).Name)
+						err = fmt.Errorf("%s is missing", e.Type().Field(i).Name)
+					}
+				}
+				if err != nil {
+					// Nack errorus message so the server gets notified that something is wrong but don't requeue the message
+					d.Nack(false, false)
+					// Send the errorus message to an error queue so it can be analyzed.
+					if err := broker.SendMessage(mq, d.CorrelationId, config.Broker.Exchange, config.Broker.RoutingError, d.Body); err != nil {
+						log.Error("faild to publish message, reason: ", err)
+					}
+					// Restart on new message
+					continue
+				}
 			}
 
-			e := reflect.ValueOf(&message).Elem()
-			for i := 0; i < e.NumField(); i++ {
-				if e.Field(i).Interface() == "" || e.Field(i).Interface() == 0 {
-					log.Errorf("%s is missing", e.Type().Field(i).Name)
-					err = fmt.Errorf("%s is missing", e.Type().Field(i).Name)
+			header, e := postgres.GetHeader(db, m.FileID)
+			if e != nil {
+				log.Error(e)
+				// Nack errorus message so the server gets notified that something is wrong but don't requeue the message
+				d.Nack(false, false)
+				// Send the errorus message to an error queue so it can be analyzed.
+				if err := broker.SendMessage(mq, d.CorrelationId, config.Broker.Exchange, config.Broker.RoutingError, d.Body); err != nil {
+					log.Error("faild to publish message, reason: ", err)
 				}
+				continue
 			}
 
-			if err == nil {
-				row := db.QueryRow(getHeader, message.FileID)
-				var hexString string
-				err = row.Scan(&hexString)
-				if err != nil {
-					log.Fatal(err)
+			// do file verfication
+			k, e := os.Open(config.Crypt4gh.KeyPath)
+			if e != nil {
+				log.Error(e)
+			}
+			defer k.Close()
+
+			key, e := keys.ReadPrivateKey(k, []byte(config.Crypt4gh.Passphrase))
+			if e != nil {
+				log.Error(e)
+			}
+
+			f, err := os.Open(filepath.Join(filepath.Clean(config.Archive.Location), m.ArchivePath))
+			if err != nil {
+				log.Error(err)
+			}
+			defer f.Close()
+
+			var buf bytes.Buffer
+			buf.Write(header)
+			var rw io.ReadWriter
+			rw = &buf
+			if _, e := io.Copy(rw, f); e != nil {
+				log.Error(e)
+			}
+
+			c4ghr, err := streaming.NewCrypt4GHReader(rw, key, nil)
+			if err != nil {
+				log.Error(err)
+			}
+			hash := sha256.New()
+			if _, err := io.Copy(hash, c4ghr); err != nil {
+				log.Error(err)
+			}
+			key = [32]byte{}
+
+			// Mark file as "COMPLETED"
+			if e := postgres.MarkCompleted(db, fmt.Sprintf("%x", hash.Sum(nil)), m.FileID); e != nil {
```

Comment:
So, as we discussed, verification will be also triggered separately, by some cron job. Do we need to touch the database every time upon this recurring verification?

Reply:
Then we need something in the message struct that tells the verifier NOT to touch the database

Reply:
Yes. But you can already add this flag to the message struct. If it's not there, it will be `null` (make it pointer). If it's there, then don't write to DB.